Fixes: Error PHP Array to string conversion at .../oc/lib/private/template/functions.php#36

zyrill opened the pull request:
See https://github.com/owncloud/core/issues/17468 for bug description.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/c96a8071-7470-46d1-99ea-15b4020381e9) was created.

karlitschek commented on lib/private/template/functions.php:
```diff
@@ -33,7 +33,7 @@
  */
 function p($string) {
-	print(OC_Util::sanitizeHTML($string));
+	implode(print(OC_Util::sanitizeHTML($string)));
```
Shouldn't this be print(implode(  ?

MorrisJobke replied:
Yep.

MorrisJobke replied:
Maybe something like:
```php
if(is_array($string)) {
   $string = implode($string);
}
print(OC_Util::sanitizeHTML($string));
```

Xenopathic replied:
:+1: to @MorrisJobke 's improvement, otherwise `implode()` will eat the input string and return null.

zyrill replied:
like it. :+1: :)

zyrill replied:
print(implode( doesn't work by the way. don't do it! will break everything... end of the world stuff. you know, hellfire raining from the skies. think sulfur burning. or that fat guy after having a lot of onions...

Xenopathic replied:
@zyrill Can you fix your PR with the suggested improvements?

zyrill replied:
yes, I will. I'm in a meeting now, give me a few hours.

zyrill commented:
@MorrisJobke anything else I need to do? I'm rather new to the github experience...

szepeviktor commented:
:+1:

szepeviktor commented:
`sanitizeHTML()` isprepared for an array.
```php
function p($string) {
	$sanitized = OC_Util::sanitizeHTML($string);
	if (is_array($sanitized)) {
		$sanitized = implode($sanitized);
	}
 	print $sanitized;
 }
```

MorrisJobke commented on lib/private/template/functions.php:
```diff
@@ -33,7 +33,11 @@
+	if (is_array($sanitized)) {
+		$sanitized = implode($sanitized);
+	}
+	print $sanitized;
```
Sorry for nitpicking: can you please add brackets after `print`? For coding guidelines compliance ;)

Thanks.

szepeviktor replied:
I was the one who left it out. Sorry.

crazybadger replied:
I've just applied this to 8.1.3 and it doesn't work!

Get an error message:

Error occurred while checking server setup

MorrisJobke replied:
> I've just applied this to 8.1.3 and it doesn't work!

It is still in development.

widac commented:
Hello, I am not familar with how this works on this page.

I just want to say I've tested this to 8.1.3 and this code works for me

```c++
function p($string) {
		$sanitized = OC_Util::sanitizeHTML($string);  
 if (is_array($sanitized)) {  
 		$sanitized = implode($sanitized);  
 	}  
 	print ($sanitized);  
```
Thanks.

VeronicaEC commented:
The above fix by @widac  doesn't work for me on 8.1.3, i got blank page accessing to owncloud webui after that.
Reverting back to:
```javascript
function p($string) {
	print(OC_Util::sanitizeHTML($string));
}
```

zyrill commented:
Alright, got the brackets in. Any more comments?

MorrisJobke commented:
> Alright, got the brackets in. Any more comments?

Can we get the 11 commits squashed into 1 because it only changes 5 lines. ;)

Then I think this is fine for review. If you need help by juggling with git feel free to ping me in IRC (#owncloud-dev on freenode)

szepeviktor commented:
`git reset HEAD~11`

Commit:
Fixes: Error PHP Array to string conversion at /share/MD0_DATA/Web/owncloud/lib/private/template/functions.php#36

See owncloud#17468 for bug description.
Incorporated changes proposed by @MorrisJobke and by @szepeviktor, see owncloud#18921

szepeviktor commented:
:hurray:

zyrill commented:
Puh, wär hätte gedacht, dass squashen so schwierig ist. Danke, @MorrisJobke. :)

MorrisJobke commented:
@zyrill Because you asked about CI - I created a local branch and PR to trigger CI: #19612

@DeepDiver1975 We need to focus on CI for forks after the release ;)

PVince81 commented:
@LukasReschke what do you think ?

LukasReschke commented:
Fair enough from my PoV. Though I'd vote only for master and no backport.

mmattel commented:
Any update on this? I also got this error on a fresh install on stable 8.2

EtienneBruines commented:
The checks are failing; that's not good. 

But I'm here to confirm that it (=changing that 1 function) did fix the error I had on fresh install (stable 8.2.0).

MorrisJobke commented:
@nickvergessen @PVince81 @LukasReschke Lets get this in :)

MorrisJobke commented:
Please review ;)

mmattel commented:
:+1:  I tested on stable 8.2 and the error did not raise again

nickvergessen commented:
I'm not sure, when it's an array it still fails to implode.
So my suggestion would be to throw a real error when it is called with an array?

szepeviktor commented:
It is meant to be called with an array:
https://github.com/zyrill/core/blob/master/lib/private/template/functions.php#L33

MorrisJobke commented:
> It is meant to be called with an array:
> https://github.com/zyrill/core/blob/master/lib/private/template/functions.php#L33

This is the weird thing ... The problem here is that it is kind of public API ... cc @DeepDiver1975

MorrisJobke commented:
> This is the weird thing ... The problem here is that it is kind of public API ... cc @DeepDiver1975

Should we change the PHPDoc - it doesn't work before anyways.

zyrill commented:
Anything needed that I can help with?

MorrisJobke commented:
> Anything needed that I can help with?

We first need a consensus ;)

PVince81 commented:
Do we actually know where the original error comes from ? Which template/app is calling this function with an array ?

MorrisJobke commented:
> Do we actually know where the original error comes from ? Which template/app is calling this function with an array ?

I activated contacts and calendar. Both don't throw the error on the admin page anymore.

I guess we should simply fix the PHPDoc to only accept `string` and then fix malicious usage of `p()` with an array as parameter. Otherwise we will get into the "black magic coding" again. @PVince81 @nickvergessen Opinions on this?

nickvergessen commented:
yes, agreeed

MorrisJobke commented:
> yes, agreeed

https://github.com/owncloud/core/pull/20679

MorrisJobke commented:
Sorry for saying this: We will close this here and need to check where this method is called in the wrong way.

If you have still this problem it would be nice to get a stack trace or at least all installed apps including their versions.

Thanks

bcutter commented:
Hi, this is definitely not fixed. I need to manually apply the fix in functions.php now sind 2015-11-19 after every single Owncloud update.

```
sudo -u www-data php /var/www/owncloud/occ status
  - installed: true
  - version: 9.0.4.1
  - versionstring: 9.0.4
  - edition:

sudo -u www-data php /var/www/owncloud/occ app:list
Enabled:
  - activity: 2.2.1
  - calendar: 1.3.2
  - comments: 0.2
  - contacts: 1.3.1.0
  - dav: 0.1.6
  - direct_menu: 0.8.1
  - documents: 0.12.1
  - federatedfilesharing: 0.1.0
  - federation: 0.0.4
  - files: 1.4.4
  - files_external: 0.5.2
  - files_locking: true
  - files_pdfviewer: 0.8.1
  - files_share_qr: 0.1
  - files_sharing: 0.9.1
  - files_texteditor: 2.1
  - files_trashbin: 0.8.0
  - files_versions: 1.2.0
  - files_videoplayer: 0.9.8
  - files_videoviewer: 0.1.3
  - firstrunwizard: 1.1
  - gallery: 14.5.0
  - logreader: 1.1.5
  - notifications: 0.2.3
  - ownbackup: 0.3.8
  - provisioning_api: 0.3.0
  - systemtags: 0.2
  - templateeditor: 0.1
  - updatenotification: 0.1.0
Disabled:
  - encryption
  - external
  - ocusagecharts
  - storagecharts2
  - storagecharts2_new
  - updater
  - user_external
  - user_ldap
  - user_otp
  - user_webdavauth
```
